File: generators/sineplasma.cpp

// Part of the ProceduralTextureMaker project.
// http://github.com/johanokl/ProceduralTextureMaker
// Released under GPLv3.
// Johan Lindqvist (johan.lindqvist@gmail.com)

#include "sineplasma.h"
#include <QtMath>
#include <QColor>
#include <cmath>

SinePlasmaTextureGenerator::SinePlasmaTextureGenerator() {
   TextureGeneratorSetting colorsetting;
   colorsetting.defaultvalue = QVariant(QColor(255, 0, 0));
   colorsetting.name = "Colour";
   colorsetting.description = "Colour mixed into the generated plasma pattern.";
   colorsetting.id = "color";
   configurables.append(colorsetting);

   TextureGeneratorSetting xoffset;
   xoffset.defaultvalue = QVariant((double)10);
   xoffset.name = "Horizontal offset (%)";
   xoffset.description = "Shifts the plasma horizontally across the texture.";
   xoffset.min = QVariant(-100);
   xoffset.max = QVariant(100);
   xoffset.id = "xoffset";
   configurables.append(xoffset);

   TextureGeneratorSetting yoffset;
   yoffset.defaultvalue = QVariant((double)10);
   yoffset.name = "Vertical offset (%)";
   yoffset.description = "Shifts the plasma vertically across the texture.";
   yoffset.min = QVariant(-100);
   yoffset.max = QVariant(100);
   yoffset.id = "yoffset";
   configurables.append(yoffset);

   TextureGeneratorSetting xfrequency;
   xfrequency.defaultvalue = QVariant((double)10);
   xfrequency.name = "Horizontal frequency";
   xfrequency.description = "Controls how often the pattern repeats along the horizontal axis.";
   xfrequency.min = QVariant(0);
   xfrequency.max = QVariant(100);
   xfrequency.group = "frequencies";
   xfrequency.id = "xfrequency";
   configurables.append(xfrequency);

   TextureGeneratorSetting yfrequency;
   yfrequency.defaultvalue = QVariant((double)10);
   yfrequency.name = "Vertical frequency";
   yfrequency.description = "Controls how often the pattern repeats along the vertical axis.";
   yfrequency.min = QVariant(0);
   yfrequency.max = QVariant(100);
   yfrequency.group = "frequencies";
   yfrequency.id = "yfrequency";
   configurables.append(yfrequency);
}
void SinePlasmaTextureGenerator::generate(QSize size, TexturePixel* destimage,
                                          const QMap<QString, TextureImagePtr>& sourceimages,
                                          const TextureNodeSettings& settings) const {
   if (!destimage || !size.isValid()) {
      return;
   }
   QColor color = settings.value("color").value<QColor>();
   double xoffset = settings.value("xoffset").toInt() * size.width() / 100;
   double yoffset = settings.value("yoffset").toInt() * size.height() / 100;
   double xfrequency = settings.value("xfrequency").toDouble() * 5 / size.width();
   double yfrequency = settings.value("yfrequency").toDouble() * 5 / size.height();

   if (sourceimages.contains(QStringLiteral("Background"))) {
      memcpy(destimage, sourceimages.value(QStringLiteral("Background"))->getData(),
             size.width() * size.height() * sizeof(TexturePixel));
   } else {
      memset(destimage, 0, size.width() * size.height() * sizeof(TexturePixel));
   }

   for (int y = 0; y < size.height(); y++) {
      for (int x = 0; x < size.width(); x++) {
         double value = 0.5 + 0.25 * qSin((x - xoffset) * xfrequency) +
                        0.25 * qSin((y - yoffset) * yfrequency);
         double negVal = 1 - value;
         int pixelPos = y * size.width() + x;
         destimage[pixelPos].r = qMin(
             qMax((int)(value * (double)color.red() + negVal * (double)destimage[pixelPos].r), 0),
             255);
         destimage[pixelPos].g = qMin(
             qMax((int)(value * (double)color.green() + negVal * (double)destimage[pixelPos].g), 0),
             255);
         destimage[pixelPos].b = qMin(
             qMax((int)(value * (double)color.blue() + negVal * (double)destimage[pixelPos].b), 0),
             255);
         destimage[pixelPos].a = qMin(
             qMax((int)(value * (double)color.alpha() + negVal * (double)destimage[pixelPos].a),
                  (int)destimage[pixelPos].a),
             255);
      }
   }
}

File: generators/sineplasma.js
const generator = {
  apiVersion: 1,
  name: "Sine plasma",
  description: "Generates a repeating two-dimensional sine-wave plasma.",
  type: "generator",
  inputs: ["Background"],

  // Each entry creates a control in the node settings panel.
  settings: [
    {
      id: "color",
      type: "color",
      name: "Colour",
      description: "Colour mixed into the generated plasma pattern.",
      default: { r: 255, g: 0, b: 0, a: 255 },
    },
    {
      id: "xoffset",
      type: "real",
      name: "Horizontal offset (%)",
      description: "Shifts the plasma horizontally across the texture.",
      default: 10,
      min: -100,
      max: 100,
    },
    {
      id: "yoffset",
      type: "real",
      name: "Vertical offset (%)",
      description: "Shifts the plasma vertically across the texture.",
      default: 10,
      min: -100,
      max: 100,
    },
    {
      id: "xfrequency",
      type: "real",
      name: "Horizontal frequency",
      description: "Controls how often the pattern repeats along the horizontal axis.",
      default: 10,
      min: 0,
      max: 100,
      group: "frequencies",
    },
    {
      id: "yfrequency",
      type: "real",
      name: "Vertical frequency",
      description: "Controls how often the pattern repeats along the vertical axis.",
      default: 10,
      min: 0,
      max: 100,
      group: "frequencies",
    },
  ],

  // This function runs whenever the application renders the node.
  generate(size, settings, output, inputs) {
    // Begin with the optional background. Without one, begin with transparency.
    const pixels = output.data;
    const background = inputs.Background?.data;
    if (background) pixels.set(background);
    else pixels.fill(0);

    // Convert the controls into pixel offsets and sine-wave frequencies.
    const width = size.width;
    const height = size.height;
    const stride = output.stride;
    // Keep the real-valued offsets fractional. Sine accepts decimal positions, so
    // there is no reason to discard the fine control offered by the settings panel.
    const xOffset = settings.xoffset * width / 100;
    const yOffset = settings.yoffset * height / 100;
    const xFrequency = settings.xfrequency * 5 / width;
    const yFrequency = settings.yfrequency * 5 / height;
    const color = settings.color;

    // A plasma value is one horizontal sine plus one vertical sine. Caching the
    // horizontal part reduces two sine calls per pixel to one per row or column.
    const horizontal = new Float64Array(width);
    for (let x = 0; x < width; ++x) {
      horizontal[x] = 0.25 * Math.sin((x - xOffset) * xFrequency);
    }

    // Step 3: use the plasma value as the selected colour's opacity, then composite
    // it over the optional background. A colour with alpha zero now has no effect.
    for (let y = 0; y < height; ++y) {
      const vertical = 0.5 + 0.25 * Math.sin((y - yOffset) * yFrequency);
      let pixel = y * stride;
      for (let x = 0; x < width; ++x, pixel += 4) {
        const value = vertical + horizontal[x];
        const plasmaAlpha = value * color.a / 255;
        if (plasmaAlpha <= 0) continue;

        const backgroundAlpha = pixels[pixel + 3] / 255;
        if (plasmaAlpha >= 1 || backgroundAlpha === 0) {
          pixels[pixel] = color.r;
          pixels[pixel + 1] = color.g;
          pixels[pixel + 2] = color.b;
          pixels[pixel + 3] = Math.round(plasmaAlpha * 255);
          continue;
        }

        const remainingBackground = 1 - plasmaAlpha;
        const resultAlpha = plasmaAlpha + backgroundAlpha * remainingBackground;
        pixels[pixel] = Math.round(
          (color.r * plasmaAlpha
            + pixels[pixel] * backgroundAlpha * remainingBackground) / resultAlpha,
        );
        pixels[pixel + 1] = Math.round(
          (color.g * plasmaAlpha
            + pixels[pixel + 1] * backgroundAlpha * remainingBackground) / resultAlpha,
        );
        pixels[pixel + 2] = Math.round(
          (color.b * plasmaAlpha
            + pixels[pixel + 2] * backgroundAlpha * remainingBackground) / resultAlpha,
        );
        pixels[pixel + 3] = Math.round(resultAlpha * 255);
      }
    }
  },
};
